plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -285,60 +507,341 @@ public boolean deleteNuageVspDevice(DeleteNuageVspDeviceCmd cmd) {
+                    validateDomains = false;
+                } else if (answer != null && answer.getSyncType() == SyncType.REGISTER) {
+                    registerNewNuageVspDevice(cmsIdConfig, nuageVspDevice.getId() + ":" + answer.getNuageVspCmsId());
                 }
-            };
-            scheduler = Executors.newScheduledThreadPool(numOfSyncThreads, threadFactory);
-            scheduler.scheduleWithFixedDelay(new NuageVspSyncTask("FLOATING_IP"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes,
-                    TimeUnit.MILLISECONDS);
-            scheduler.scheduleWithFixedDelay(new NuageVspSyncTask("ENTERPRISE_NTWK_MACRO"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes,
-                    TimeUnit.MILLISECONDS);
-            scheduler
-                    .scheduleWithFixedDelay(new NuageVspSyncTask("ENTERPRISE"), ONE_MINUTE_MULTIPLIER * 15, ONE_MINUTE_MULTIPLIER * syncUpIntervalInMinutes, TimeUnit.MILLISECONDS);
-        } else {
-            s_logger.warn("NuageVsp configuration for syncWorkers=" + numOfSyncThreads + " syncInterval=" + syncUpIntervalInMinutes
-                    + " could not be read properly. So, check if the properties are configured properly in global properties");
+            }
+        }
+
+        if (validateDomains) {
+            auditDomainsOnVsp(host, true, false);
         }
     }
 
-    public class NuageVspSyncTask implements Runnable {
+    private boolean auditDomainsOnVsp(HostVO host, boolean add, boolean remove) {
+        List<NuageVspDeviceVO> nuageVspDevices = _nuageVspDao.listByHost(host.getId());
+        if (!CollectionUtils.isEmpty(nuageVspDevices)) {
+            _hostDao.loadDetails(host);
+            List<DomainVO> allDomains = _domainDao.listAll();
+            for (DomainVO domain : allDomains) {
+                try {
+                    SyncDomainCommand cmd = new SyncDomainCommand(domain.getUuid(), domain.getName(), domain.getPath(), add, remove);
+                    SyncDomainAnswer answer = (SyncDomainAnswer) _agentMgr.easySend(host.getId(), cmd);
+                    return answer.getSuccess();
+                } catch (Exception e) {
+                    s_logger.warn(e.getMessage());
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
 
-        private String nuageVspEntity;
+    private String findNuageVspCmsIdForDevice(long deviceId, ConfigurationVO cmsIdConfig) {
+        String configValue = cmsIdConfig.getValue();
+        if (StringUtils.isNotBlank(configValue)) {
+            String[] configuredNuageVspDevices = configValue.split(";");
+            for (String configuredNuageVspDevice : configuredNuageVspDevices) {
+                if (configuredNuageVspDevice.startsWith(deviceId + ":")) {
+                    return configuredNuageVspDevice.split(":")[1];
```

Comment:
This is a specific value that's being set from within the plugin itself, and this is in a specific format like <ID>:<UUID>. I'll make the code fail safe though.